README.md

```diff
@@ -337,6 +339,190 @@ PING 10.0.0.1 (10.0.0.1) 56(84) bytes of data.
+```shell
+KDIR = <linux kernel top directory for VMs>
+```
+
+Save the change and build `vwifi`:
+```shell
+make
+```
+
+On success, we expect that `vwifi.ko` will show in the top directory of `vwifi`.
+
+The cfg80211 API and net device API in the kernel have changed along with the new kernel version, and forgive us for not testing for all of them.
+### Build Rootfs by Buildroot
+We use `buildroot` to build our rootfs, you can either choose your desired tool for building rootfs.
+
+Get the `builtroot` from [https://buildroot.org](https://buildroot.org).
+
+Enter the top directory of the `buildroot` source and use the following command to generate the first-phase configuration file:
+```shell
+make qemu_x86_64_defconfig
+```
+
+The `qemu_x86_64_defconfig` configuration will not fit the testing environment, so we need to further configure it:
+```shell
+make menuconfig
+```
+
+We need a filesystem overlay for putting our `vwifi` kernel module and some user program's configuration file into the VM, so choose a place for the filesystem overlay (we recommand creating an empty directory for it).
+
+Also, we need network applications like `iw`, `hostapd` and `wpa_supplicant`.
+```
+System configuration ---> Run a getty (login prompt) after boot ---> TTY port ---> ttyS0
+System configuration ---> Root filesystem overlay directories ---> <the place you wnat>
+Kernel ---> Linux Kernel ---> no
+Target packages ---> Networking applications ---> hostapd
+Target packages ---> Networking applications ---> iw
+Target packages ---> Networking applications ---> wpa_supplicant
+Filesystem images ---> ext2/3/4 root filesystem
+```
+
+After finishing the configuration, we can put the needed kernel modules and configuration files into the filesystem overlay directory:
+```shell
+cp <linux-top-dir>/net/wireless/cfg80211.ko \
+<vwifi-top-dir>/vwifi.ko \
+<vwifi-top-dir>/scripts/hostapd.conf \
+<vwifi-top-dir>/scripts/wpa_supplicant.conf \
+<overlay-dir>
+```
+
+Then start building rootfs:
+```shell
+make
+```
+This may take a long time, please be patient.
+
+The output image will be `output/images/rootfs.ext2`. We need three images for three VMs, so we simply copy this file into three:
+```shell
+cp output/images/rootfs.ext2 output/images/rootfs2.ext2
+cp output/images/rootfs.ext2 output/images/rootfs3.ext2
+```
+### Setup Host Network Device
+We need three `tap` devices and each of them must be attached to a `bridge` device. Please note, creating `tap` and `bridge` devices needs privilege permission.
+
+Creating three `tap` devices:
+```shell
+sudo ip tuntap add mode tap tap0
+sudo ip tuntap add mode tap tap1
+sudo ip tuntap add mode tap tap2
+```
+
+Creating `bridge` device:
+```shell
+sudo ip link add name br0 type bridge
+```
+
+Attach three `tap` devices on `bridge` device:
+```shell
+sudo ip link set tap0 master br0
+sudo ip link set tap1 master br0
+sudo ip link set tap2 master br0
+```
+
+### Start VM with Qemu
```

Comment:
How about integrating into CI pipeline?

Reply:
The CI yields the following error: `ioctl(TUNSETIFF): Operation not permitted`.

It seems that only the privileged users can access the tun/tap devices. So we couldn't add these lines into the CI pipeline.